File: cpp/13_Roman_to_Integer.cpp
/*
Given a roman numeral, convert it to an integer.
Input is guaranteed to be within the range from 1 to 3999.
*/

#include <iostream>
#include <vector>

using namespace std;

string symbol = "IVXLCDM";
vector<int> symbolValue = {1, 5, 10, 50, 100, 500, 1000};

class Solution {

    public:
        int romanToInt(string s) {
            int sum = 0;
            int prevIdx = -1;
            for(int i = 0; i < s.length(); i++) {
                int curInx = symbol.find(s[i], 0);
                if(prevIdx != -1 && prevIdx < curInx) {
                    sum -= 2 * symbolValue[prevIdx];
                }
                sum += symbolValue[curInx];
                prevIdx = curInx;
            }
            return sum;
        }
    
};

int main() {
    Solution sol;
    cout << to_string(sol.romanToInt("DCCC") == 800) << endl;
    cout << to_string(sol.romanToInt("XCIX") == 99) << endl;
    cout << to_string(sol.romanToInt("LXXXIX") == 89) << endl;
    cout << to_string(sol.romanToInt("LXXIV") == 74) << endl;
    return 0;
}

File: py/13_Roman_to_Integer.py
"""
Given a roman numeral, convert it to an integer.
Input is guaranteed to be within the range from 1 to 3999.
"""

class Solution:
    def romanToInt(self, s: str) -> int:
        conv = {'M': 1000, 'D': 500, 'C': 100, 'L': 50, 'X': 10, 'V': 5, 'I': 1}
        num, prev = 0, None
        for ch in s:
            value = conv[ch]
            num += value
            if prev and prev < value:
                num -= 2 * prev
            prev = value
        return num
